File: javascript/nodes/prompt_m3u_file.js
import { Next } from '../ambler.js';
import { createInterface } from 'readline';

const defaultUtils = {
    readLine: () => new Promise((resolve) => {
        const rl = createInterface({ input: process.stdin });
        rl.once('line', (line) => {
            rl.close();
            resolve(line.trim());
        });
    }),
};

export function promptM3UFile(edges, utils = defaultUtils) {
    return async (state) => {
        console.log('Please enter the path to your M3U file:');
        const m3uFilePath = await utils.readLine();

        if (!m3uFilePath) {
            return new Next(promptM3UFile(edges, utils), state);
        }

        return new Next(edges.onCheck, { ...state, m3uFilePath });
    };
}

File: ruby/nodes/prompt_m3u_file.rb
# frozen_string_literal: true

require_relative '../lib/ambler'

module PromptM3uFile
  DEFAULT_UTILS = {
    read_line: lambda {
      print 'Please enter the path to your M3U file: '
      $stdin.gets.to_s.chomp
    }
  }.freeze

  def self.node(edges, utils = DEFAULT_UTILS)
    self_node = ->(state) {
      m3u_file_path = utils[:read_line].call

      if m3u_file_path.empty?
        return Ambler::Next.new(self_node, state)
      end

      Ambler::Next.new(edges[:on_check], state.merge(m3u_file_path: File.expand_path(m3u_file_path)))
    }
    self_node
  end
end
